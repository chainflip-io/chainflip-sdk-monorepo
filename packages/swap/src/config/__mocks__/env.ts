export default {
  RPC_NODE_HTTP_URL: 'http://rpc-node.test',
  START_HTTP_SERVICE: false,
  START_PROCESSOR: false,
  SWAPPING_APP_PORT: 8080,
  BITCOIN_DEPOSIT_TRACKER_URL: '',
  RPC_BROKER_HTTPS_URL: 'https://rpc-broker.test',
  RPC_NODE_WSS_URL: 'ws://rpc-node.test',
  CHAINFLIP_NETWORK: 'perseverance',
  QUOTE_TIMEOUT: 1000,
  NODE_ENV: 'test',
  CHAINALYSIS_API_KEY: undefined,
  INGEST_GATEWAY_URL: 'https://ingest-gateway.test',
  PROCESSOR_BATCH_SIZE: 50,
  PROCESSOR_TRANSACTION_TIMEOUT: 10_000,
  REDIS_URL: 'redis://localhost:6379',
  MAINTENANCE_MODE: false,
  FULLY_DISABLED_INTERNAL_ASSETS: new Set(),
  DISABLED_DEPOSIT_INTERNAL_ASSETS: new Set(),
  DISABLED_DESTINATION_INTERNAL_ASSETS: new Set(),
  MAX_CHANNELS_OPEN_PER_ADDRESS: 10,
  DISABLE_QUOTING: false,
  SOLANA_RPC_HTTP_URL: 'http://solana-rpc.test',
  BITCOIN_RPC_HTTP_URL: 'http://bitcoin-rpc.test',
  STABLE_COIN_SLIPPAGE_MIN_PRICE: 0.995,
  DCA_CHUNK_INTERVAL_BLOCKS: 2,
  QUOTER_BALANCE_TOLERANCE_PERCENT: 10,
  QUOTING_BASE_SLIPPAGE: {},
  QUOTER_USE_AUGMENT: true,
  BROKER_COMMISSION_BPS: 0,
  ON_CHAIN_DEFAULT_NETWORK_FEE_BPS: 5n,
  ON_CHAIN_STABLECOIN_NETWORK_FEE_BPS: 1n,
  MINIMUM_NETWORK_FEE_USDC: 500_000n,
};
