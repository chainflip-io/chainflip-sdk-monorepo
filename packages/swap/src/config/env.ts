import { z } from 'zod';

const httpUrl = z.string().url().startsWith('http');
const wsUrl = z.string().url().startsWith('ws');

const optionalBoolean = z
  .string()
  .optional()
  .transform((value) => value?.toUpperCase() === 'TRUE');

const optionalNumber = (defaultValue: number) =>
  z
    .string()
    .optional()
    .transform((n) => Number(n) || defaultValue);

const chainflipNetwork = z.enum([
  'backspin',
  'sisyphos',
  'perseverance',
  'mainnet',
]);

const nodeEnv = z.enum(['development', 'production', 'test']);

export default z
  .object({
    RPC_NODE_HTTP_URL: httpUrl,
    START_HTTP_SERVICE: optionalBoolean,
    START_PROCESSOR: optionalBoolean,
    SWAPPING_APP_PORT: optionalNumber(8080),
    BITCOIN_DEPOSIT_TRACKER_URL: httpUrl.optional(),
    RPC_BROKER_HTTPS_URL: httpUrl,
    RPC_NODE_WSS_URL: wsUrl,
    CHAINFLIP_NETWORK: chainflipNetwork,
    QUOTE_TIMEOUT: optionalNumber(1000),
    NODE_ENV: nodeEnv.default('production'),
    CHAINALYSIS_API_KEY: z.string().optional(),
    INGEST_GATEWAY_URL: httpUrl,
    PROCESSOR_BATCH_SIZE: optionalNumber(50),
    PROCESSOR_TRANSACTION_TIMEOUT: optionalNumber(10_000),
  })
  // eslint-disable-next-line n/no-process-env
  .parse(process.env);
