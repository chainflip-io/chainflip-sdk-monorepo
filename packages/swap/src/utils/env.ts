import { ChainflipAsset } from '@chainflip/utils/chainflip';
import ServiceError from './ServiceError.js';
import env from '../config/env.js';

export const isLocalnet = () => env.CHAINFLIP_NETWORK === 'backspin';

export const assertRouteEnabled = ({
  srcAsset,
  destAsset,
}: {
  srcAsset: ChainflipAsset;
  destAsset: ChainflipAsset;
}) => {
  if (
    env.FULLY_DISABLED_INTERNAL_ASSETS.has(srcAsset) ||
    env.DISABLED_DEPOSIT_INTERNAL_ASSETS.has(srcAsset)
  ) {
    throw ServiceError.unavailable(`Asset ${srcAsset} is disabled`);
  }

  if (
    env.FULLY_DISABLED_INTERNAL_ASSETS.has(destAsset) ||
    env.DISABLED_DESTINATION_INTERNAL_ASSETS.has(destAsset)
  ) {
    throw ServiceError.unavailable(`Asset ${destAsset} is disabled`);
  }
};
