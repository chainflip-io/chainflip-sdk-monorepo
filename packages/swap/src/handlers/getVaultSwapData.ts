import { HttpClient } from '@chainflip/rpc';
import { z } from 'zod';
import { getParameterEncodingRequestSchema } from '@/shared/broker';
import { getInternalAsset } from '@/shared/enums';
import { transformKeysToCamelCase } from '@/shared/objects';
import env from '../config/env';
import isDisallowedSwap from '../utils/isDisallowedSwap';
import logger from '../utils/logger';
import { validateSwapAmount } from '../utils/rpc';
import ServiceError from '../utils/ServiceError';

const client = new HttpClient(env.RPC_BROKER_HTTPS_URL);

export const getVaultSwapData = async (
  input: z.output<ReturnType<typeof getParameterEncodingRequestSchema>>,
) => {
  logger.info('Fetching vault swap data', input);

  if (
    await isDisallowedSwap(
      input.destAddress,
      input.srcAddress,
      input.fillOrKillParams?.refund_address,
    )
  ) {
    logger.info('Blocked address found for vault swap data', input);
    throw ServiceError.internalError('Failed to get vault swap data, please try again later');
  }

  const srcAsset = getInternalAsset(input.srcAsset);
  if (env.DISABLED_INTERNAL_ASSETS.includes(srcAsset)) {
    throw ServiceError.unavailable(`Asset ${srcAsset} is disabled`);
  }

  const destAsset = getInternalAsset(input.destAsset);
  if (env.DISABLED_INTERNAL_ASSETS.includes(destAsset)) {
    throw ServiceError.unavailable(`Asset ${destAsset} is disabled`);
  }

  const result = await validateSwapAmount(srcAsset, BigInt(input.amount));

  if (!result.success) throw ServiceError.badRequest(result.reason);

  const response = await client.sendRequest(
    'broker_request_swap_parameter_encoding',
    input.srcAsset,
    input.destAsset,
    input.destAddress,
    0, // default broker does not support commission
    input.extraParams,
    input.ccmParams,
    input.maxBoostFeeBps,
    undefined, // default broker does not support affiliates
    input.dcaParams,
  );

  logger.info('Vault swap data fetched', response);

  return transformKeysToCamelCase(response);
};
