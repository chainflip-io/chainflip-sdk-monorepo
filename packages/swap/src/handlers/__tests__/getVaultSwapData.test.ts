import { vi, describe, it, beforeEach, afterEach, expect } from 'vitest';
import { environment, mockRpcResponse } from '@/shared/tests/fixtures';
import env from '@/swap/config/env';
import disallowSwap from '../../utils/disallowSwap';
import { getVaultSwapData } from '../getVaultSwapData';

vi.mock('@/shared/broker', () => ({
  requestSwapParameterEncoding: vi.fn(),
}));

vi.mock('../../utils/disallowSwap', () => ({
  default: vi.fn().mockResolvedValue(false),
}));

describe(getVaultSwapData, () => {
  let oldEnv: typeof env;

  beforeEach(async () => {
    oldEnv = structuredClone(env);
  });

  afterEach(() => {
    Object.assign(env, oldEnv);
  });

  it('calls the rpc with the correct params', async () => {
    const postSpy = mockRpcResponse((url, data: any) => {
      if (data.method === 'cf_environment') {
        return Promise.resolve({ data: environment({ maxSwapAmount: '0x10000000000' }) });
      }

      if (data.method === 'broker_request_swap_parameter_encoding') {
        return Promise.resolve({
          data: {
            id: '1',
            jsonrpc: '2.0',
            result: {
              chain: 'Bitcoin',
              nulldata_payload:
                '0x0003656623d865425c0a4955ef7e7a39d09f58554d0800000000000000000000000000000000000001000200000100',
              deposit_address: 'bcrt1pmrhjpvq2w7cgesrcrvuhqw6n6j487l6uc7tmwtx9jen7ezesunhqllvzxx',
            },
          },
        });
      }

      throw new Error(`unexpected axios call to ${url}: ${JSON.stringify(data)}`);
    });

    const result = await getVaultSwapData({
      srcAsset: { asset: 'BTC', chain: 'Bitcoin' },
      destAsset: { asset: 'ETH', chain: 'Ethereum' },
      destAddress: '0xe983fD1798689eee00c0Fb77e79B8f372DF41060',
      amount: 175000000n,
      commissionBps: 15,
      extraParams: {
        chain: 'Bitcoin',
        min_output_amount: '0xfc6f7c40458122964d0000000',
        retry_duration: 500,
      },
      fillOrKillParams: {
        retry_duration: 500,
        refund_address: '0xa56A6be23b6Cf39D9448FF6e897C29c41c8fbDFF',
        min_price: '0x152d02c7e14af680000000000000000000000000000000000000',
      },
      ccmParams: {
        gas_budget: '0x75bcd15',
        message: '0xdeadc0de',
        cf_parameters: undefined,
      },
      maxBoostFeeBps: 35,
      affiliates: [
        {
          account: 'cFHyJEHEQ1YkT9xuFnxnPWVkihpYEGjBg4WbF6vCPtSPQoE8n',
          bps: 10,
        },
      ],
      dcaParams: {
        number_of_chunks: 100,
        chunk_interval: 5,
      },
    });

    expect(postSpy).toHaveBeenCalledWith('https://rpc-broker.test', [
      {
        id: expect.any(String),
        jsonrpc: '2.0',
        method: 'broker_request_swap_parameter_encoding',
        params: [
          {
            asset: 'BTC',
            chain: 'Bitcoin',
          },
          {
            asset: 'ETH',
            chain: 'Ethereum',
          },
          '0xe983fD1798689eee00c0Fb77e79B8f372DF41060',
          0,
          {
            chain: 'Bitcoin',
            min_output_amount: '0xfc6f7c40458122964d0000000',
            retry_duration: 500,
          },
          {
            gas_budget: '0x75bcd15',
            message: '0xdeadc0de',
          },
          35,
          null,
          {
            chunk_interval: 5,
            number_of_chunks: 100,
          },
        ],
      },
    ]);
    expect(result).toEqual({
      chain: 'Bitcoin',
      nulldataPayload:
        '0x0003656623d865425c0a4955ef7e7a39d09f58554d0800000000000000000000000000000000000001000200000100',
      depositAddress: 'bcrt1pmrhjpvq2w7cgesrcrvuhqw6n6j487l6uc7tmwtx9jen7ezesunhqllvzxx',
    });
  });

  it('rejects sanctioned addresses', async () => {
    vi.mocked(disallowSwap).mockResolvedValueOnce(true);

    await expect(
      getVaultSwapData({
        srcAsset: { asset: 'BTC', chain: 'Bitcoin' },
        destAsset: { asset: 'ETH', chain: 'Ethereum' },
        srcAddress: 'bc1qqwykx04uenc842d3sf50cjtehtj9tenugk808w',
        destAddress: '0xe983fD1798689eee00c0Fb77e79B8f372DF41060',
        amount: 175000000n,
        commissionBps: 15,
        extraParams: {
          chain: 'Bitcoin',
          min_output_amount: '0xfc6f7c40458122964d0000000',
          retry_duration: 500,
        },
        fillOrKillParams: {
          retry_duration: 500,
          refund_address: 'bc1qv6nkwfd0l7gjucvccgd4k4qcea488cfkg7zjcp',
          min_price: '0x152d02c7e14af680000000000000000000000000000000000000',
        },
      }),
    ).rejects.toThrow('Failed to get vault swap data, please try again later');
    expect(vi.mocked(disallowSwap).mock.calls).toMatchSnapshot();
  });

  it('rejects if source asset is disabled', async () => {
    env.DISABLED_INTERNAL_ASSETS = ['Flip', 'Btc'];

    await expect(
      getVaultSwapData({
        srcAsset: { asset: 'BTC', chain: 'Bitcoin' },
        destAsset: { asset: 'ETH', chain: 'Ethereum' },
        destAddress: '0xe983fD1798689eee00c0Fb77e79B8f372DF41060',
        amount: 175000000n,
        commissionBps: 15,
        extraParams: {
          chain: 'Bitcoin',
          min_output_amount: '0xfc6f7c40458122964d0000000',
          retry_duration: 500,
        },
        fillOrKillParams: {
          retry_duration: 500,
          refund_address: '0xa56A6be23b6Cf39D9448FF6e897C29c41c8fbDFF',
          min_price: '0x152d02c7e14af680000000000000000000000000000000000000',
        },
      }),
    ).rejects.toThrow('Asset Btc is disabled');
  });

  it('rejects if destination asset is disabled', async () => {
    env.DISABLED_INTERNAL_ASSETS = ['Btc', 'Dot'];

    await expect(
      getVaultSwapData({
        srcAsset: { asset: 'ETH', chain: 'Ethereum' },
        destAsset: { asset: 'DOT', chain: 'Polkadot' },
        destAddress: '1yMmfLti1k3huRQM2c47WugwonQMqTvQ2GUFxnU7Pcs7xPo',
        amount: 175000000n,
        commissionBps: 15,
        extraParams: {
          chain: 'Bitcoin',
          min_output_amount: '0xfc6f7c40458122964d0000000',
          retry_duration: 500,
        },
        fillOrKillParams: {
          retry_duration: 500,
          refund_address: '0xa56A6be23b6Cf39D9448FF6e897C29c41c8fbDFF',
          min_price: '0x152d02c7e14af680000000000000000000000000000000000000',
        },
      }),
    ).rejects.toThrow('Asset Dot is disabled');
  });
});
