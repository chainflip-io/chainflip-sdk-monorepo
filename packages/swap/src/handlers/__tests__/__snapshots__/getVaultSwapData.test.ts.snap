// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`encodeVaultSwapData > rejects sanctioned addresses 1`] = `
[
  [
    "0xe983fD1798689eee00c0Fb77e79B8f372DF41060",
    undefined,
    "0xa56A6be23b6Cf39D9448FF6e897C29c41c8fbDFF",
  ],
  [
    "0xe983fD1798689eee00c0Fb77e79B8f372DF41060",
    "bc1qqwykx04uenc842d3sf50cjtehtj9tenugk808w",
    "bc1qv6nkwfd0l7gjucvccgd4k4qcea488cfkg7zjcp",
  ],
]
`;
